dubbo-remoting/dubbo-remoting-api/src/main/java/org/apache/dubbo/remoting/api/Connection.java

```diff
@@ -192,10 +192,8 @@ public boolean isAvailable() {
         if (channel != null && channel.isActive()) {
             return true;
         }
-        synchronized (this) {
```

Comment:
Under multi-threading, the function of `sync` here is to let those threads that have not acquired locks wait. When these waiting threads are released from waiting, it can be ensured that the `bootstrap.connect()` operation must have been initiated (this operation is acquired locked thread).

I think the `init` property here can be replaced with a `volatile boolean `type

#9381

Reply:
Volatile variables may not work well. In any case, judgment is required. If it is only volatile, concurrency may occur when judging and modifying if.

Reply:
yes, judgment and assignment operations are originally needed, I mean code like the following

        ```
if(!init){
            synchronized (this) {
                if (!init) {
                    connect();
                    init = true;
                }
            }
        }
```

Reply:
Atomic variables might be better?

Reply:
![image](https://user-images.githubusercontent.com/13098309/164725902-5e406b1b-4b3c-4197-926b-a4075fafd29c.png)

`sync` has added when creating `promise`, so is it enough that just a `cas` here  ?

Reply:
Assume that thread 1 passes the cas , thread2 skips the cas, no matter thread1 or thread2 creates a promise, both threads will wait for it complete.

Reply:
agree with @guohao